packages/native/src/resolvers/ffi_resolver.rs

```diff
@@ -17,19 +17,34 @@ use super::{
     resolution_context::FFIUriResolutionContext, uri_package_or_wrapper::FFIUriPackageOrWrapper,
 };
 
-pub trait FFIUriResolver: Send + Sync + Debug {
+pub trait IFFIUriResolver: Send + Sync + Debug {
     fn try_resolve_uri(
         &self,
         uri: Arc<FFIUri>,
         invoker: Arc<FFIInvoker>,
         resolution_context: Arc<FFIUriResolutionContext>,
-    ) -> Result<Box<dyn FFIUriPackageOrWrapper>, FFIError>;
+    ) -> Result<Arc<FFIUriPackageOrWrapper>, FFIError>;
 }
 
 #[derive(Debug)]
-pub struct UriResolverWrapping(pub Box<dyn FFIUriResolver>);
+pub struct FFIUriResolver(pub Box<dyn IFFIUriResolver>);
 
-impl UriResolver for UriResolverWrapping {
+impl FFIUriResolver {
+    pub fn new(uri_resolver: Box<dyn IFFIUriResolver>) -> Self {
+        Self(uri_resolver)
+    }
+
+    pub fn try_resolve_uri(
+        &self,
+        uri: Arc<FFIUri>,
+        invoker: Arc<FFIInvoker>,
+        resolution_context: Arc<FFIUriResolutionContext>,
+    ) -> Result<Arc<FFIUriPackageOrWrapper>, FFIError> {
+        self.0.try_resolve_uri(uri, invoker, resolution_context)
+    }
+}
+
+impl UriResolver for FFIUriResolver {
```

Comment:
I am a bit confused, why are doing this but in `Wrapper` we're doing `impl IFFIWrapper for Arc<dyn Wrapper>` ? I was trying to do some tests but I haven't been able to understand how I should use the API.

What I was trying to do is that, if we have this `MockInvoker`:
```
#[derive(Debug)]
pub struct MockResolver;

impl UriResolver for MockResolver {
    fn try_resolve_uri(
        &self,
        uri: &Uri,
        _: Arc<dyn Invoker>,
        _: Arc<Mutex<UriResolutionContext>>,
    ) -> Result<UriPackageOrWrapper, Error> {
        if uri.to_string() == *"wrap://wrap/mock" {
            Ok(get_mock_uri_package_or_wrapper())
        } else {
            Err(Error::ResolutionError("Not Found".to_string()))
        }
    }
}

pub fn get_mock_resolver() -> Arc<dyn UriResolver> {
    Arc::new(MockResolver {})
}
```
I was expecting to be able to do something like:
```
let resolver = Box::new(get_mock_resolver());
FFIUriResolver(resolver);
```
but this doesn't work. What am I missing?

Reply:
**Short answer**: make your mock wrappers implement `IFFIWrapper` not `Wrapper`. You're testing the API from an `FFIClient` user perspective. The consumer of Rust's FFI, cannot use `Wrapper` as it is not exposed; nor is it FFI compatible. The consumer would import and implement `IFFIWrapper`.

**Longer answer**:

That said, when using uri resolution, a `UriResolver`'s `try_resolve_uri` will return a `UriPackageOrWrapper` which is a struct that contains an `Arc<dyn Wrapper>`; and no further information. In this case, we need a way to transform an `Arc<dyn Wrapper>` into an `FFIWrapper` so that it can be returned by an `FFIUriResolver`. The easiest way to achieve this, and probably the cleanest, is to implement `IFFIWrapper` for `Arc<dyn Wrapper>`, that way we can:

```rust
let boxed_wrapper: Box<dyn IFFIWrapper> = Box::new(arc_wrapper);
let wrapper: FFIWrapper = FFIWrapper::new(boxed_wrapper)
```

The downside to this, is having to wrap the object in a `Box` which makes a heap allocation; but in this particular case I don't think it's a bad thing.